File: schedule.js
// schedule.js
const axios = require('axios');
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let cache = {};
const CACHE_TTL = 60;

async function fetchConnections(from, to, date, time) {
    const key = `${from}:${to}:${date}:${time}`;
    if (cache[key]) {
        const entry = cache[key];
        if ((Date.now() - entry.timestamp) / 1000 < CACHE_TTL) {
            return entry.data;
        }
    }
    const url = 'https://transport.opendata.ch/v1/connections';
    const params = { from, to };
    if (date) params.date = date;
    if (time) params.time = time;
    try {
        const response = await axios.get(url, { params, timeout: 10000 });
        const data = response.data.connections || [];
        cache[key] = { data, timestamp: Date.now() };
        return data;
    } catch (e) {
        console.error('Error fetching data:', e.message);
        return [];
    }
}

function formatConnection(c) {
    return {
        departure: c.from.departure || '',
        departureStation: c.from.station.name || '',
        arrival: c.to.arrival || '',
        arrivalStation: c.to.station.name || '',
        duration: c.duration || '',
        changes: c.transfers || 0
    };
}

function display(connections, limit = 10, sortBy = 'departure') {
    if (!connections || !connections.length) {
        console.log('No connections found.');
        return;
    }
    let formatted = connections.map(c => formatConnection(c));
    if (sortBy === 'departure') {
        formatted.sort((a, b) => a.departure.localeCompare(b.departure));
    } else if (sortBy === 'duration') {
        formatted.sort((a, b) => a.duration.localeCompare(b.duration));
    } else if (sortBy === 'changes') {
        formatted.sort((a, b) => a.changes - b.changes);
    }
    console.log(`\nFound ${formatted.length} connections.`);
    console.log('-'.repeat(80));
    console.log(`${'#'.padEnd(3)} ${'Departure'.padEnd(20)} ${'Arrival'.padEnd(20)} ${'Duration'.padEnd(10)} ${'Changes'.padEnd(8)}`);
    for (let i = 0; i < Math.min(formatted.length, limit); i++) {
        const c = formatted[i];
        const depTime = c.departure ? c.departure.slice(11, 16) : '?';
        const arrTime = c.arrival ? c.arrival.slice(11, 16) : '?';
        const depSt = c.departureStation.slice(0, 10);
        const arrSt = c.arrivalStation.slice(0, 10);
        console.log(`${String(i+1).padEnd(3)} ${depTime} ${depSt.padEnd(10)} ${arrTime} ${arrSt.padEnd(10)} ${c.duration.padEnd(10)} ${c.changes}`);
    }
    if (formatted.length > limit) {
        console.log(`... and ${formatted.length - limit} more`);
    }
}

function exportJSON(connections, filename) {
    fs.writeFileSync(filename, JSON.stringify(connections, null, 2));
    console.log(`Exported to ${filename}`);
}

function exportCSV(connections, filename) {
    const lines = ['Departure,Departure Station,Arrival,Arrival Station,Duration,Changes'];
    connections.forEach(c => {
        const f = formatConnection(c);
        lines.push(`${f.departure},"${f.departureStation}",${f.arrival},"${f.arrivalStation}",${f.duration},${f.changes}`);
    });
    fs.writeFileSync(filename, lines.join('\n'));
    console.log(`Exported to ${filename}`);
}

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

async function interactive() {
    console.log('=== Train Schedule ===');
    const from = await ask('From station: ');
    if (!from.trim()) { console.log('Station required.'); rl.close(); return; }
    const to = await ask('To station: ');
    if (!to.trim()) { console.log('Station required.'); rl.close(); return; }
    let date = await ask('Date (YYYY-MM-DD, leave blank for today): ');
    if (!date.trim()) date = new Date().toISOString().slice(0,10);
    let time = await ask('Time (HH:MM, leave blank for now): ');
    if (!time.trim()) time = new Date().toTimeString().slice(0,5);
    const conns = await fetchConnections(from.trim(), to.trim(), date.trim(), time.trim());
    display(conns);
    if (conns.length) {
        const exportAns = await ask('\nExport results? (y/n): ');
        if (exportAns.toLowerCase() === 'y') {
            const fmt = await ask('Format (json/csv): ');
            let fname = await ask('Filename (default: schedule.' + fmt + '): ');
            if (!fname.trim()) fname = 'schedule.' + fmt;
            if (fmt === 'json') exportJSON(conns, fname);
            else if (fmt === 'csv') exportCSV(conns, fname);
            else console.log('Unknown format.');
        }
    }
    rl.close();
}

function cli() {
    const args = require('minimist')(process.argv.slice(2));
    if (!args.from || !args.to) {
        console.error('--from and --to are required.');
        process.exit(1);
    }
    const from = args.from;
    const to = args.to;
    const date = args.date || '';
    const time = args.time || '';
    const limit = args.limit || 10;
    const sortBy = args.sort || 'departure';
    fetchConnections(from, to, date, time).then(conns => {
        display(conns, limit, sortBy);
        if (args.export && conns.length) {
            const fname = args.output || `schedule.${args.export}`;
            if (args.export === 'json') exportJSON(conns, fname);
            else if (args.export === 'csv') exportCSV(conns, fname);
        }
    });
}

if (process.argv.length > 2) {
    cli();
} else {
    interactive();
}

File: schedule.rb
# schedule.rb
require 'net/http'
require 'json'
require 'csv'
require 'time'
require 'optparse'

$cache = {}
CACHE_TTL = 60

def fetch_connections(from, to, date = nil, time = nil)
  key = "#{from}:#{to}:#{date}:#{time}"
  if $cache.key?(key)
    entry = $cache[key]
    if Time.now - entry[:timestamp] < CACHE_TTL
      return entry[:data]
    end
  end
  uri = URI('https://transport.opendata.ch/v1/connections')
  params = { from: from, to: to }
  params[:date] = date if date
  params[:time] = time if time
  uri.query = URI.encode_www_form(params)
  response = Net::HTTP.get_response(uri)
  if response.is_a?(Net::HTTPSuccess)
    data = JSON.parse(response.body)
    connections = data['connections'] || []
    $cache[key] = { data: connections, timestamp: Time.now }
    connections
  else
    puts "Error: #{response.code}"
    []
  end
rescue => e
  puts "Error: #{e.message}"
  []
end

def format_connection(c)
  {
    departure: c['from']['departure'],
    departure_station: c['from']['station']['name'],
    arrival: c['to']['arrival'],
    arrival_station: c['to']['station']['name'],
    duration: c['duration'],
    changes: c['transfers'] || 0
  }
end

def display(connections, limit = 10, sort_by = 'departure')
  if connections.empty?
    puts 'No connections found.'
    return
  end
  formatted = connections.map { |c| format_connection(c) }
  case sort_by
  when 'duration'
    formatted.sort_by! { |c| c[:duration] }
  when 'changes'
    formatted.sort_by! { |c| c[:changes] }
  else
    formatted.sort_by! { |c| c[:departure] }
  end
  puts "\nFound #{formatted.size} connections."
  puts '-' * 80
  puts "%-3s %-20s %-20s %-10s %-8s" % ['#', 'Departure', 'Arrival', 'Duration', 'Changes']
  formatted.first(limit).each_with_index do |c, i|
    dep_time = c[:departure] ? c[:departure][11..15] : '?'
    arr_time = c[:arrival] ? c[:arrival][11..15] : '?'
    dep_st = c[:departure_station][0..9]
    arr_st = c[:arrival_station][0..9]
    puts "%-3d %s %-10s %s %-10s %-10s %-8d" % [i+1, dep_time, dep_st, arr_time, arr_st, c[:duration], c[:changes]]
  end
  if formatted.size > limit
    puts "... and #{formatted.size - limit} more"
  end
end

def export_json(connections, filename)
  File.write(filename, JSON.pretty_generate(connections))
  puts "Exported to #{filename}"
end

def export_csv(connections, filename)
  CSV.open(filename, 'w') do |csv|
    csv << ['Departure', 'Departure Station', 'Arrival', 'Arrival Station', 'Duration', 'Changes']
    connections.each do |c|
      csv << [c['from']['departure'], c['from']['station']['name'], c['to']['arrival'], c['to']['station']['name'], c['duration'], c['transfers'] || 0]
    end
  end
  puts "Exported to #{filename}"
end

def interactive
  puts "=== Train Schedule ==="
  print "From station: "
  from = gets.chomp.strip
  if from.empty?
    puts "Station required."
    return
  end
  print "To station: "
  to = gets.chomp.strip
  if to.empty?
    puts "Station required."
    return
  end
  print "Date (YYYY-MM-DD, leave blank for today): "
  date = gets.chomp.strip
  date = Time.now.strftime('%Y-%m-%d') if date.empty?
  print "Time (HH:MM, leave blank for now): "
  time = gets.chomp.strip
  time = Time.now.strftime('%H:%M') if time.empty?
  conns = fetch_connections(from, to, date, time)
  display(conns)
  if !conns.empty?
    print "\nExport results? (y/n): "
    export_ans = gets.chomp.strip.downcase
    if export_ans == 'y'
      print "Format (json/csv): "
      fmt = gets.chomp.strip.downcase
      print "Filename (default: schedule.#{fmt}): "
      fname = gets.chomp.strip
      fname = "schedule.#{fmt}" if fname.empty?
      if fmt == 'json'
        export_json(conns, fname)
      elsif fmt == 'csv'
        export_csv(conns, fname)
      else
        puts "Unknown format."
      end
    end
  end
end

def cli
  options = {}
  OptionParser.new do |opts|
    opts.on('--from STATION', 'Departure station') { |v| options[:from] = v }
    opts.on('--to STATION', 'Destination station') { |v| options[:to] = v }
    opts.on('--date DATE', 'Date (YYYY-MM-DD)') { |v| options[:date] = v }
    opts.on('--time TIME', 'Time (HH:MM)') { |v| options[:time] = v }
    opts.on('--limit N', Integer, 'Max connections') { |v| options[:limit] = v }
    opts.on('--sort FIELD', 'Sort field') { |v| options[:sort] = v }
    opts.on('--export FORMAT', 'Export format (json/csv)') { |v| options[:export] = v }
    opts.on('--output FILE', 'Output filename') { |v| options[:output] = v }
  end.parse!
  if options[:from].nil? || options[:to].nil?
    puts "--from and --to are required."
    exit 1
  end
  conns = fetch_connections(options[:from], options[:to], options[:date], options[:time])
  display(conns, options[:limit] || 10, options[:sort] || 'departure')
  if options[:export] && !conns.empty?
    fname = options[:output] || "schedule.#{options[:export]}"
    if options[:export] == 'json'
      export_json(conns, fname)
    elsif options[:export] == 'csv'
      export_csv(conns, fname)
    end
  end
end

if ARGV.empty?
  interactive
else
  cli
end
